gobblin-api/src/main/java/org/apache/gobblin/stream/ControlMessageInjector.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.stream;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.metadata.GlobalMetadata;
+import org.apache.gobblin.records.ControlMessageHandler;
+import org.apache.gobblin.records.RecordStreamProcessor;
+import org.apache.gobblin.records.RecordStreamWithMetadata;
+
+import io.reactivex.Flowable;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.Setter;
+
+/**
+ * A {@link RecordStreamProcessor} that inspects an input record and outputs control messages before, after, or around
+ * the input record
+ * @param <SI>
+ * @param <DI>
+ */
+public abstract class ControlMessageInjector<SI, DI> implements Closeable,
+        RecordStreamProcessor<SI, SI, DI, DI> {
+
+  @Setter(AccessLevel.PROTECTED)
+  @Getter(AccessLevel.PROTECTED)
+  private GlobalMetadata<SI> inputGlobalMetadata;
+
+  /**
+   * Initialize this {@link ControlMessageInjector}.
+   *
+   * @param workUnitState a {@link WorkUnitState} object carrying configuration properties
+   * @return an initialized {@link ControlMessageInjector} instance
+   */
+  public ControlMessageInjector<SI, DI> init(WorkUnitState workUnitState) {
```

Comment:
Can this be `protected` instead? Same with all methods except `processStream`.

Reply:
Done.